draft-ietf-mls-protocol.md

```diff
@@ -1686,24 +1626,70 @@ decryption.
 +-+-+-+-+---------...---+
 ~~~~~
 
-The Additional Authenticated Data (AAD) input to the encryption
-contains an object of the following form, with the values used to
-identify the key and nonce:
+## Masking Sender Data
+
+The "sender data" used to look up the key for the content encryption is masked
+with the output of the PRF specified in the group's ciphersuite. Specifically,
+the encoded sender data is XORed with the PRF output of the first block of
+ciphertext under the key `sender_data_key`, and stored in the
+`masked_sender_data` field of MLSCiphertext. If the ciphertext is less than one
+block long, it's padded on the right with zeroes.
+
+Before being masked, the sender data is encoded as an object of the following
+form:
+
+~~~~~
+struct {
+    uint32 sender;
+    uint32 generation;
+    opaque reuse_guard[4];
+} MLSSenderData;
+~~~~~
+
+MLSSenderData.sender is assumed to be a `member` sender type.  When constructing
+an MLSSenderData from a Sender object, the sender MUST verify Sender.sender_type
+is `member` and use Sender.sender for MLSSenderData.sender.
+
+The `reuse_guard` field contains a fresh random value used to avoid nonce reuse
+in the case of state loss or corruption, as described in {{content-encryption}}.
+
+When parsing a SenderData struct as part of message decryption, the
+recipient MUST verify that the sender field represents an occupied
+leaf in the ratchet tree.  In particular, the sender index value
+MUST be less than the number of leaves in the tree.
+
+### AES-Based Masking
+
+When the ciphersuite specifies AES for masking metadata, this is accomplished by
+encrypting the first block of ciphertext with AES in ECB mode. In pseudocode:
+
+~~~
+mask = AES-ECB(sender_data_key, ciphertext[0..15])
+~~~
+
+## Message Authentication
+
+The `auth_tag` field of MLSCiphertext is produced by calling the AEAD function
+specified by the ciphersuite for a second time. This time, the key is
+`authentication_key`, and the nonce is the unmasked sender data (possibly padded
+on the right with zeroes to match AEAD.nonce\_length). The plaintext is empty
+and the Additional Authenticated Data (AAD) input contains an object of the
+following form:
 
 ~~~~~
 struct {
     opaque group_id<0..255>;
     uint64 epoch;
     ContentType content_type;
     opaque authenticated_data<0..2^32-1>;
-    opaque sender_data_nonce<0..255>;
-    opaque encrypted_sender_data<0..255>;
-} MLSCiphertextContentAAD;
+    opaque ciphertext<0..2^32-1>;
+} MLSCiphertextAAD;
```

Comment:
You've removed the AAD from the main content encryption, which seems wrong.  it seems like we should have the unencrypted sender data feed in the ciphertext AAD, and the ciphertext (or the derived nonce) in the sender data AAD.  Something like:

```
      /  group_id            \
      |  epoch               |--+
   +--|  content_type        /  |
   |  |  sender_data         <--+
   |  \  authenticated_data  \__|
   +-->  content             /

struct {
    opaque group_id<0..255>;
    uint64 epoch;
    ContentType content_type;
    opaque authenticated_data<0..2^32-1>; // <---
    opaque ciphertext<0..2^32-1>;
} SenderDataAAD;

struct {
    opaque group_id<0..255>;
    uint64 epoch;
    ContentType content_type;
    SenderData sender_data;
    opaque authenticated_data<0..2^32-1>; // <--- 
} ContentAAD;
```

Reply:
The ciphertext's auth tag is truncated so there's no point in having AAD on the content. All AAD is against the sender data.